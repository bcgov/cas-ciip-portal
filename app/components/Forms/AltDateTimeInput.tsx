// https://github.com/rjsf-team/react-jsonschema-form/blob/v1.8.1/src/components/widgets/AltDateTimeWidget.js

import React from "react";
import PropTypes from "prop-types";
import AltDateInput from "./AltDateInput";
import { WidgetProps } from "react-jsonschema-form";

function AltDateTimeInput(props: WidgetProps) {
  const { AltDateInput } = props.registry.widgets;
  return <AltDateInput time {...props} />;
}

if (process.env.NODE_ENV !== "production") {
  AltDateTimeInput.propTypes = {
    schema: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    value: PropTypes.string,
    required: PropTypes.bool,
    onChange: PropTypes.func,
    options: PropTypes.object,
  };
}

AltDateTimeInput.defaultProps = {
  ...AltDateInput.defaultProps,
  time: true,
};

export default AltDateTimeInput;