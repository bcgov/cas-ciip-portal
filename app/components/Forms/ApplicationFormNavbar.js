import React from 'react';
import {Nav} from 'react-bootstrap';
import {createFragmentContainer} from 'react-relay';
import Link from 'next/link';

const ApplicationFormNavbarComponent = props => {
  const {wizard, application} = props;
  return (
    <>
      <Nav justify variant="tabs" defaultActiveKey="/home">
        {wizard.edges.map(({node}) => (
          <Nav.Item key={node.formJsonByFormId.id}>
            <Link
              passHref
              href={{
                pathname: '/ciip-application',
                query: {
                  formId: node.formJsonByFormId.id,
                  applicationId: application.id
                }
              }}
            >
              <Nav.Link>{node.formJsonByFormId.name}</Nav.Link>
            </Link>
          </Nav.Item>
        ))}
      </Nav>
    </>
  );
};

export default createFragmentContainer(ApplicationFormNavbarComponent, {});
