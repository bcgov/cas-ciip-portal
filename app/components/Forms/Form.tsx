import JsonSchemaForm from 'react-jsonschema-form';
import {toErrorList} from 'react-jsonschema-form/lib/validate';
/**
 * Extends the Form class from 'react-jsonchema-form'
 * to enable validation on blur for all fields
 */
export default class Form<T> extends JsonSchemaForm<T> {
  constructor(props) {
    super(props);
    const superOnBlur = this.onBlur;

    this.onBlur = (id, value) => {
      // Id is 'root_path_to_field'
      const fieldPath = id.split('_');
      // Error schema is an object whose shape matches the form data
      const {formData, errorSchema} = this.state as any;
      const {errorSchema: newErrorSchema} = this.validate(formData);

      // We only want to retrieve the errors for the blurred field in newErrorSchema
      // as this.validate will validate all fields
      let blurredFieldErrorSchema = newErrorSchema;
      for (let i = 1; i < fieldPath.length; i++) {
        blurredFieldErrorSchema = blurredFieldErrorSchema[fieldPath[i]];
      }

      // Although we only want to add the errors from the blurred field,
      // we still want to keep the errors from previously blurred fields
      const mergedErrorSchema = {...errorSchema};

      // Find the object that will contain blurredFieldErrorSchema
      let blurredFieldParentSchema = mergedErrorSchema;
      for (let i = 1; i < fieldPath.length - 1; i++) {
        if (blurredFieldParentSchema[fieldPath[i]] === undefined)
          blurredFieldParentSchema[fieldPath[i]] = {};

        // Make a copy of the part of the error schema that was blurred.
        // This ensures immutability of the error schema, and proper re-rendering of the form
        blurredFieldParentSchema[fieldPath[i]] = {
          ...blurredFieldParentSchema[fieldPath[i]]
        };

        // Go one level down in the error schema tree
        blurredFieldParentSchema = blurredFieldParentSchema[fieldPath[i]];
      }

      blurredFieldParentSchema[
        fieldPath[fieldPath.length - 1]
      ] = blurredFieldErrorSchema;

      this.setState(
        {
          errors: toErrorList(mergedErrorSchema),
          errorSchema: mergedErrorSchema
        },
        () => {
          superOnBlur(id, value);
        }
      );
    };
  }
}
