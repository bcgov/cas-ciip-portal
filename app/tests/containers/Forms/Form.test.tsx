import React from 'react';
import {render, shallow} from 'enzyme';
import {FormComponent} from '../../../containers/Forms/Form';
import adminForm from '../../../../schema/data/portal/form_json/administration.json';
import emissionForm from '../../../../schema/data/portal/form_json/emission.json';
import fuelForm from '../../../../schema/data/portal/form_json/fuel.json';
import electricityAndHeatForm from '../../../../schema/data/portal/form_json/electricity_and_heat.json';
import productionForm from '../../../../schema/data/portal/form_json/production.json';

describe('Form', () => {
  it('should match the snapshot with the administration form', async () => {
    /** In node_modules/react-jsonSchema-form/lib/components/widgets/RadioWidget there is a 'var name = Math.random().toString();'
        The comment above this line says: // Generating a unique field name to identify this set of radio buttons.
        -> This is why we are patching Math.random() to return zero for this test.
    */
    Math.random = () => 0;
    const r = render(
      <FormComponent
        query={{
          ' $refType': 'Form_query',
          ' $fragmentRefs': {
            FuelFields_query: true,
            ProductionFields_query: true
          },
          result: {
            formResult: {operator: {name: 'Test operator'}},
            formJsonByFormId: {
              formJson: adminForm,
              ciipApplicationWizardByFormId: {
                formPosition: 0
              }
            }
          }
        }}
      />
    );
    expect(r).toMatchSnapshot();
    expect(r.find('.form-submit a').length).toBe(1);
    expect(r.find('Button').length).toBe(1);
    expect(r.find('Button').text()).toBe('Continue');
  });

  it('should match the snapshot with the fuel form', async () => {
    // TODO: figure out how to test fragment container under fuel
    const TestRenderer = () => (
      <FormComponent
        query={{
          ' $refType': 'Form_query',
          ' $fragmentRefs': {
            FuelFields_query: true,
            ProductionFields_query: true
          },
          result: {
            formResult: [
              {
                fuelType: 'C/D Waste - Plastic',
                quantity: 4,
                fuelUnits: 't',
                methodology: 'wci 1.0'
              }
            ],
            formJsonByFormId: {
              formJson: fuelForm
            }
          }
        }}
      />
    );
    // Shallow used as 'render' tries to continue farther down because of the ArrayFieldTemplate
    const r = shallow(<TestRenderer />);
    expect(r).toMatchSnapshot();
  });

  it('should match the snapshot with the emission form', async () => {
    // Shallow used as 'render' tries to continue farther down because of the ArrayFieldTemplate
    const r = render(
      <FormComponent
        query={{
          ' $refType': 'Form_query',
          ' $fragmentRefs': {
            FuelFields_query: true,
            ProductionFields_query: true
          },
          result: {
            formResult: {
              sourceTypes: [
                {
                  sourceType: 'Flaring',
                  gases: [
                    {
                      gasType: 'CH4',
                      GWP: ' x 25 = ',
                      annualCO2e: 1625,
                      annualEmission: 65,
                      gasDescription: 'gassy'
                    },
                    {
                      gasType: 'C02',
                      GWP: ' x 2 = ',
                      annualCO2e: 25,
                      annualEmission: 5,
                      gasDescription: 'not as gassy'
                    }
                  ]
                }
              ]
            },
            formJsonByFormId: {
              formJson: emissionForm,
              ciipApplicationWizardByFormId: {
                formPosition: 1
              }
            }
          }
        }}
      />
    );
    expect(r).toMatchSnapshot();
    expect(r.find('.form-submit a').length).toBe(1);
    expect(r.find('Button').length).toBe(2);
    expect(r.find('Button.btn-secondary').text()).toBe('Back');
  });

  it('should match the snapshot with the electricity and heat form', async () => {
    const r = render(
      <FormComponent
        query={{
          ' $refType': 'Form_query',
          ' $fragmentRefs': {
            FuelFields_query: true,
            ProductionFields_query: true
          },
          result: {
            formResult: {heat: {sold: 81}, electricity: {sold: 81}},
            formJsonByFormId: {
              formJson: electricityAndHeatForm
            }
          }
        }}
      />
    );
    expect(r).toMatchSnapshot();
  });

  it('should match the snapshot with the production form', async () => {
    // Shallow used as 'render' tries to continue farther down the product-units fragment
    const r = shallow(
      <FormComponent
        query={{
          ' $refType': 'Form_query',
          ' $fragmentRefs': {
            FuelFields_query: true,
            ProductionFields_query: true
          },
          result: {
            formResult: [
              {
                product: 'Dehydration',
                comments: 'Saepe quis aliquid e',
                quantity: 84,
                productUnits: 'kl',
                associatedEmissions: 42
              }
            ],
            formJsonByFormId: {
              formJson: productionForm
            }
          }
        }}
      />
    );
    expect(r).toMatchSnapshot();
  });
});
