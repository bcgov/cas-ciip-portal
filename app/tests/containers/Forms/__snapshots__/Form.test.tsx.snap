// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot with the administration form 1`] = `
<form
  class="rjsf"
>
  <div
    class="col-12"
  >
    <div
      class="form-row"
    >
      <div
        class="col-12"
      >
        <legend>
          Administrative Information
        </legend>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-12"
          >
            <legend>
              Operator Information
            </legend>
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operator_name"
            >
              Legal name
            </label>
            <input
              class="form-control"
              id="root_operator_name"
              label="Legal name"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operator_tradeName"
            >
              Trade name
            </label>
            <input
              class="form-control"
              id="root_operator_tradeName"
              label="Trade name"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operator_naics"
            >
              NAICS code
            </label>
            <input
              class="form-control"
              id="root_operator_naics"
              label="NAICS code"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operator_bcCorporateRegistryNumber"
            >
              BC Corporate Registry number
            </label>
            <input
              class="form-control"
              id="root_operator_bcCorporateRegistryNumber"
              label="BC Corporate Registry number"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operator_duns"
            >
              DUNS number
            </label>
            <input
              class="form-control"
              id="root_operator_duns"
              label="DUNS number"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="col-12"
          >
            <div
              class="form-row"
            >
              <div
                class="col-12"
              >
                <legend>
                  Operator Mailing Address
                </legend>
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operator_mailingAddress_streetAddress"
                >
                  Street address
                </label>
                <input
                  class="form-control"
                  id="root_operator_mailingAddress_streetAddress"
                  label="Street address"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operator_mailingAddress_city"
                >
                  City
                </label>
                <input
                  class="form-control"
                  id="root_operator_mailingAddress_city"
                  label="City"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operator_mailingAddress_postalCode"
                >
                  Postal Code
                </label>
                <input
                  class="form-control"
                  id="root_operator_mailingAddress_postalCode"
                  label="Postal Code"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operator_mailingAddress_province"
                >
                  Province or Territory
                </label>
                <select
                  class="form-control"
                  id="root_operator_mailingAddress_province"
                >
                  <option
                    selected=""
                    value=""
                  />
                  <option
                    value="Alberta"
                  >
                    Alberta
                  </option>
                  <option
                    value="British Columbia"
                  >
                    British Columbia
                  </option>
                  <option
                    value="Manitoba"
                  >
                    Manitoba
                  </option>
                  <option
                    value="New Brunswick"
                  >
                    New Brunswick
                  </option>
                  <option
                    value="Newfoundland and Labrador"
                  >
                    Newfoundland and Labrador
                  </option>
                  <option
                    value="Northwest Territories"
                  >
                    Northwest Territories
                  </option>
                  <option
                    value="Nova Scotia"
                  >
                    Nova Scotia
                  </option>
                  <option
                    value="Nunavut"
                  >
                    Nunavut
                  </option>
                  <option
                    value="Ontario"
                  >
                    Ontario
                  </option>
                  <option
                    value="Prince Edward Island"
                  >
                    Prince Edward Island
                  </option>
                  <option
                    value="Quebec"
                  >
                    Quebec
                  </option>
                  <option
                    value="Saskatchewan"
                  >
                    Saskatchewan
                  </option>
                  <option
                    value="Yukon"
                  >
                    Yukon
                  </option>
                </select>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-12"
          >
            <legend>
              Operational Representative
            </legend>
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_firstName"
            >
              First Name
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_firstName"
              label="First Name"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_lastName"
            >
              Last Name
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_lastName"
              label="Last Name"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_position"
            >
              Position/Title
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_position"
              label="Position/Title"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_email"
            >
              Email Address
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_email"
              label="Email Address"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_phone"
            >
              Telephone Number
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_phone"
              label="Telephone Number"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_operationalRepresentative_fax"
            >
              Fax Number
            </label>
            <input
              class="form-control"
              id="root_operationalRepresentative_fax"
              label="Fax Number"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="col-12"
          >
            <div
              class="form-row"
            >
              <div
                class="col-12"
              >
                <legend>
                  Operational Representative Mailing Address
                </legend>
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operationalRepresentative_mailingAddress_streetAddress"
                >
                  Street address
                </label>
                <input
                  class="form-control"
                  id="root_operationalRepresentative_mailingAddress_streetAddress"
                  label="Street address"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operationalRepresentative_mailingAddress_city"
                >
                  City
                </label>
                <input
                  class="form-control"
                  id="root_operationalRepresentative_mailingAddress_city"
                  label="City"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operationalRepresentative_mailingAddress_postalCode"
                >
                  Postal Code
                </label>
                <input
                  class="form-control"
                  id="root_operationalRepresentative_mailingAddress_postalCode"
                  label="Postal Code"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_operationalRepresentative_mailingAddress_province"
                >
                  Province or Territory
                </label>
                <select
                  class="form-control"
                  id="root_operationalRepresentative_mailingAddress_province"
                >
                  <option
                    selected=""
                    value=""
                  />
                  <option
                    value="Alberta"
                  >
                    Alberta
                  </option>
                  <option
                    value="British Columbia"
                  >
                    British Columbia
                  </option>
                  <option
                    value="Manitoba"
                  >
                    Manitoba
                  </option>
                  <option
                    value="New Brunswick"
                  >
                    New Brunswick
                  </option>
                  <option
                    value="Newfoundland and Labrador"
                  >
                    Newfoundland and Labrador
                  </option>
                  <option
                    value="Northwest Territories"
                  >
                    Northwest Territories
                  </option>
                  <option
                    value="Nova Scotia"
                  >
                    Nova Scotia
                  </option>
                  <option
                    value="Nunavut"
                  >
                    Nunavut
                  </option>
                  <option
                    value="Ontario"
                  >
                    Ontario
                  </option>
                  <option
                    value="Prince Edward Island"
                  >
                    Prince Edward Island
                  </option>
                  <option
                    value="Quebec"
                  >
                    Quebec
                  </option>
                  <option
                    value="Saskatchewan"
                  >
                    Saskatchewan
                  </option>
                  <option
                    value="Yukon"
                  >
                    Yukon
                  </option>
                </select>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-12"
          >
            <legend>
              Facility Information
            </legend>
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_facility_facilityName"
            >
              Facility Name
            </label>
            <input
              class="form-control"
              id="root_facility_facilityName"
              label="Facility Name"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_facility_facilityType"
            >
              Facility Type
            </label>
            <select
              class="form-control"
              id="root_facility_facilityType"
            >
              <option
                selected=""
                value=""
              />
              <option
                value="IF_a"
              >
                IF_a
              </option>
              <option
                value="IF_b"
              >
                IF_b
              </option>
              <option
                value="L_c"
              >
                L_c
              </option>
              <option
                value="SFO"
              >
                SFO
              </option>
              <option
                value="LFO"
              >
                LFO
              </option>
              <option
                value="EIO"
              >
                EIO
              </option>
            </select>
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_facility_bcghgid"
            >
              BCGHG ID Number
            </label>
            <input
              class="form-control"
              id="root_facility_bcghgid"
              label="BCGHG ID Number"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_facility_naics"
            >
              NAICS Code
            </label>
            <input
              class="form-control"
              id="root_facility_naics"
              label="NAICS Code"
              placeholder=""
              type="text"
              value=""
            />
          </div>
          <div
            class="col-12"
          >
            <div
              class="form-row"
            >
              <div
                class="col-12"
              >
                <legend>
                  Facility Mailing Address
                </legend>
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_facility_mailingAddress_streetAddress"
                >
                  Street address
                </label>
                <input
                  class="form-control"
                  id="root_facility_mailingAddress_streetAddress"
                  label="Street address"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_facility_mailingAddress_city"
                >
                  City
                </label>
                <input
                  class="form-control"
                  id="root_facility_mailingAddress_city"
                  label="City"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_facility_mailingAddress_postalCode"
                >
                  Postal Code
                </label>
                <input
                  class="form-control"
                  id="root_facility_mailingAddress_postalCode"
                  label="Postal Code"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group field field-string form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
              >
                <label
                  class="form-label"
                  for="root_facility_mailingAddress_province"
                >
                  Province or Territory
                </label>
                <select
                  class="form-control"
                  id="root_facility_mailingAddress_province"
                >
                  <option
                    selected=""
                    value=""
                  />
                  <option
                    value="Alberta"
                  >
                    Alberta
                  </option>
                  <option
                    value="British Columbia"
                  >
                    British Columbia
                  </option>
                  <option
                    value="Manitoba"
                  >
                    Manitoba
                  </option>
                  <option
                    value="New Brunswick"
                  >
                    New Brunswick
                  </option>
                  <option
                    value="Newfoundland and Labrador"
                  >
                    Newfoundland and Labrador
                  </option>
                  <option
                    value="Northwest Territories"
                  >
                    Northwest Territories
                  </option>
                  <option
                    value="Nova Scotia"
                  >
                    Nova Scotia
                  </option>
                  <option
                    value="Nunavut"
                  >
                    Nunavut
                  </option>
                  <option
                    value="Ontario"
                  >
                    Ontario
                  </option>
                  <option
                    value="Prince Edward Island"
                  >
                    Prince Edward Island
                  </option>
                  <option
                    value="Quebec"
                  >
                    Quebec
                  </option>
                  <option
                    value="Saskatchewan"
                  >
                    Saskatchewan
                  </option>
                  <option
                    value="Yukon"
                  >
                    Yukon
                  </option>
                </select>
              </div>
            </div>
          </div>
          <div
            class="form-group field field-boolean form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_facility_isFacilityLocationDifferent"
            >
              Is the facility location different from the mailing address?
            </label>
            <div
              class="field-radio-group"
              id="root_facility_isFacilityLocationDifferent"
            >
              <div
                class="radio "
              >
                <label>
                  <span>
                    <input
                      name="0"
                      type="radio"
                      value="true"
                    />
                    <span>
                       Yes
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="radio "
              >
                <label>
                  <span>
                    <input
                      name="0"
                      type="radio"
                      value="false"
                    />
                    <span>
                       No
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style="text-align:right"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Form should match the snapshot with the electricity and heat form 1`] = `
<form
  class="rjsf"
>
  <div
    class="col-12"
  >
    <div
      class="form-row"
    >
      <div
        class="col-12"
      >
        <legend>
          Emissions
        </legend>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-md-12 col-12"
          >
            <div
              class="col-12"
            >
              <div
                class="form-row"
              >
                <div
                  class="col-12"
                >
                  <legend />
                </div>
                <div
                  class="form-group field field-string  hidden-title form-group col-xl-12 col-lg-12 col-md-12 col-sm-12 col-12"
                >
                  <label
                    class="form-label"
                    for="root_sourceTypes_0_sourceTypeName"
                  >
                    sourceTypeName
                  </label>
                  <div
                    class="row"
                    style="margin:20px 0 20px 0"
                  >
                    <div
                      class="col-md-4 col-12"
                    >
                      <h5
                        class="jsx-2806087631"
                        style="text-align:left"
                      >
                        Stationary Combustion
                      </h5>
                    </div>
                    <div
                      class="col-md-3 col-12"
                    >
                      <h6
                        class="jsx-2806087631"
                      >
                        Tonnes
                      </h6>
                    </div>
                    <div
                      class="col-md-2 col-12"
                    >
                      <h6
                        class="jsx-2806087631"
                      >
                        GWP
                      </h6>
                    </div>
                    <div
                      class="col-md-3 col-12"
                    >
                      <h6
                        class="jsx-2806087631"
                      >
                        Tonnes (CO2e)
                      </h6>
                    </div>
                  </div>
                </div>
                <div
                  class="col-12"
                >
                  <div
                    class="form-row"
                  >
                    <div
                      class="col-md-12 col-12"
                    >
                      <div
                        class="col-12"
                      >
                        <div
                          class="form-row"
                        >
                          <div
                            class="col-md-12 col-12"
                          >
                            <div
                              class="form-row"
                            >
                              <div
                                class="col-md-4"
                              >
                                CO2 from non-bio 
                                <br
                                  class="jsx-4064704399"
                                />
                                <div
                                  class="col-md-10"
                                  style="margin:0;padding:0;color:#888;line-height:17px"
                                >
                                  <small
                                    class="jsx-4064704399"
                                  />
                                </div>
                              </div>
                              <div
                                class="col-md-3"
                              >
                                <input
                                  class="form-control"
                                  type="number"
                                  value="20"
                                />
                              </div>
                              <div
                                class="col-md-2"
                                style="text-align:center;margin-top:5px"
                              >
                                <ul
                                  class="jsx-4064704399 gwp"
                                >
                                  <li
                                    class="jsx-4064704399"
                                  >
                                    X
                                  </li>
                                  <li
                                    class="jsx-4064704399"
                                  >
                                    250
                                  </li>
                                  <li
                                    class="jsx-4064704399"
                                  >
                                    =
                                  </li>
                                </ul>
                              </div>
                              <div
                                class="col-md-3"
                              >
                                <input
                                  class="form-control"
                                  disabled=""
                                  type="number"
                                  value="100"
                                />
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <hr
                    class="jsx-529402721"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <hr
          class="jsx-529402721"
        />
      </div>
    </div>
  </div>
  <div
    style="text-align:right"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Form should match the snapshot with the electricity and heat form 2`] = `
<form
  class="rjsf"
>
  <div
    class="col-12"
  >
    <div
      class="form-row"
    >
      <div
        class="col-12"
      >
        <legend>
          Electricity and Heat Information
        </legend>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-12"
          >
            <legend>
              Electricity
            </legend>
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_electricity_purchased"
            >
              Purchased
            </label>
            <input
              class="form-control"
              id="root_electricity_purchased"
              label="Purchased"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_electricity_sold"
            >
              Sold
            </label>
            <input
              class="form-control"
              id="root_electricity_sold"
              label="Sold"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_electricity_consumedOnSite"
            >
              Consumed On-Site
            </label>
            <input
              class="form-control"
              id="root_electricity_consumedOnSite"
              label="Consumed On-Site"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_electricity_generatedOnSite"
            >
              Generated On-Site
            </label>
            <input
              class="form-control"
              id="root_electricity_generatedOnSite"
              label="Generated On-Site"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="col-12"
      >
        <div
          class="form-row"
        >
          <div
            class="col-12"
          >
            <legend>
              Heat
            </legend>
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_heat_purchased"
            >
              Purchased
            </label>
            <input
              class="form-control"
              id="root_heat_purchased"
              label="Purchased"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_heat_sold"
            >
              Sold
            </label>
            <input
              class="form-control"
              id="root_heat_sold"
              label="Sold"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-4 col-lg-4 col-md-4 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_heat_consumedOnSite"
            >
              Consumed On-Site
            </label>
            <input
              class="form-control"
              id="root_heat_consumedOnSite"
              label="Consumed On-Site"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
          <div
            class="form-group field field-number form-group col-xl-6 col-lg-6 col-md-6 col-sm-12 col-12"
          >
            <label
              class="form-label"
              for="root_heat_generatedOnSite"
            >
              Generated On-Site
            </label>
            <input
              class="form-control"
              id="root_heat_generatedOnSite"
              label="Generated On-Site"
              placeholder=""
              step="any"
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    style="text-align:right"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Form should match the snapshot with the fuel form 1`] = `
<form
  class="rjsf"
>
  <div
    class="col-12"
  >
    <div
      class="form-row"
    >
      <div
        class="col-12"
      >
        <div
          class="jsx-529402721 add-button-container"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Add a fuel
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    style="text-align:right"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Form should match the snapshot with the production form 1`] = `
<form
  class="rjsf"
>
  <div
    class="col-12"
  >
    <div
      class="form-row"
    >
      <div
        class="col-12"
      >
        <div
          class="jsx-529402721 add-button-container"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Add a Product
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    style="text-align:right"
  >
    <button
      class="btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
